bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -0,0 +1,1004 @@
+                        .setValue(event.getProperties().getMemoryUsage(isSHCClassic));
+                onDeviceStateChanged(bridgeDevice);
+            }
+        }
+    }
+
+    @Override
+    public void onEvent(final String msg) {
+        logger.trace("onEvent called. Msg: {}", msg);
+
+        try {
+            final Optional<EventDTO> eventOptional = parseEvent(msg);
+            if (eventOptional.isPresent()) {
+                EventDTO event = eventOptional.get();
+                switch (event.getType()) {
+                    case BaseEventDTO.TYPE_STATE_CHANGED:
+                    case BaseEventDTO.TYPE_BUTTON_PRESSED:
+                        handleStateChangedEvent(event, deviceStructMan);
+                        break;
+                    case BaseEventDTO.TYPE_DISCONNECT:
+                        logger.debug("Websocket disconnected.");
+                        scheduleRestartClient(true);
+                        break;
+                    case BaseEventDTO.TYPE_CONFIGURATION_CHANGED:
+                        handleConfigurationChangedEvent(event);
+                        break;
+                    case BaseEventDTO.TYPE_CONTROLLER_CONNECTIVITY_CHANGED:
+                        handleControllerConnectivityChangedEvent(event, deviceStructMan);
+                        break;
+                    case BaseEventDTO.TYPE_NEW_MESSAGE_RECEIVED:
+                    case BaseEventDTO.TYPE_MESSAGE_CREATED:
+                        final Optional<MessageEventDTO> messageEvent = gson.fromJson(msg, MessageEventDTO.class);
+                        if (messageEvent.isPresent()) {
+                            handleNewMessageReceivedEvent(Objects.requireNonNull(messageEvent.get()), deviceStructMan);
+                        }
+                        break;
+                    case BaseEventDTO.TYPE_MESSAGE_DELETED:
+                        handleMessageDeletedEvent(event, deviceStructMan);
+                        break;
+                    default:
+                        logger.debug("Unsupported event type {}.", event.getType());
+                        break;
+                }
+            }
+        } catch (IOException | ApiException | AuthenticationException | RuntimeException e) {
+            logger.debug("Error with Event: {}", e.getMessage(), e);
+            handleClientException(e);
+        }
+    }
+
+    @Override
+    public void onError(final Throwable cause) {
+        if (cause instanceof Exception) {
+            handleClientException((Exception) cause);
+        }
+    }
+
+    /**
+     * Handles the event that occurs, when the state of a device (like reachability) or a capability (like a temperature
+     * value) has changed.
+     *
+     * @param event event
+     */
+    private void handleStateChangedEvent(final EventDTO event, @Nullable final DeviceStructureManager deviceStructMan)
+            throws ApiException, IOException, AuthenticationException {
+
+        if (deviceStructMan != null) {
```

Comment:
It's impossible that DeviceStructureManager is NULL... When this happens, the whole binding isn't working. That is an example what I mean with useless non-null-checks. In other Java applications I would throw a hard exception when this happens, but this isn't allowed at OpenHAB...

Reply:
Where did you see that this is not allowed? If it really can't happen, you can throw a `RuntimeException` in this case. In this method you could throw `IllegalArgumentException`.

Reply:
At other code reviews, netatmo binding or withings binding. It was said that I have to write a debug message instead of throwing an exception... I really don't understand the special OpenHAB error handling. Exceptions shouldn't be use and no logging levels other than debug should be used... Also the changes within this code-review. We can not differ anymore between debug messages or occurred error situations...

Reply:
If you can point to a concrete review comment, I could have a look. In general throwing exceptions is fine, even preferred, so error can be handled at the appropriate place. Uncaught RuntimeExceptions should only be used in impossible/unexpected situations in order to reveal bugs, as it will make the binding stop working. But IMHO still better in some situations instead of writing very defensive code.

Reply:
When I think about it again, it doesn't help to throw an Exception to avoid the non-null messages, because the member variable has still to be declared as nullable (with all consequences...). It's actually the same discussion with the private method approach or local variables. I have just created a private method which can handle a nullable DeviceStructureManager (just extracted the null handling to a method instead of declaring a second nullable local variable to do the null check).

Reply:
and I think it is a good approach to extract that to the called method as said in the other threads (no confusing variables, small methods, the flow of the original method kept clean, etc.).